src/transport/message_handler.rs

```diff
@@ -0,0 +1,188 @@
+use crate::events::Event;
+use crate::transport::msgs::{LSPSMessage, RawLSPSMessage, LSPS_MESSAGE_TYPE};
+use crate::transport::protocol::LSPS0MessageHandler;
+
+use bitcoin::secp256k1::PublicKey;
+use lightning::ln::features::{InitFeatures, NodeFeatures};
+use lightning::ln::peer_handler::CustomMessageHandler;
+use lightning::ln::wire::CustomMessageReader;
+use lightning::log_info;
+use lightning::sign::EntropySource;
+use lightning::util::logger::Logger;
+use lightning::util::ser::Readable;
+use std::collections::HashMap;
+use std::convert::TryFrom;
+use std::io;
+use std::ops::Deref;
+use std::sync::{Arc, Mutex};
+
+const LSPS_FEATURE_BIT: usize = 729;
+
+/// A trait used to implement a specific LSPS protocol
+/// The messages the protocol uses need to be able to be mapped
+/// from and into LSPSMessages.
+pub(crate) trait ProtocolMessageHandler {
+	type ProtocolMessage: TryFrom<LSPSMessage> + Into<LSPSMessage>;
+	const PROTOCOL_NUMBER: Option<u16>;
+
+	fn handle_message(
+		&self, message: Self::ProtocolMessage, counterparty_node_id: &PublicKey,
+	) -> Result<(), lightning::ln::msgs::LightningError>;
+}
+
+/// Configuration for the LSPManager
+/// Allows end-user to configure the behavior of the client
+pub struct LSPConfig {
+	/// whether or not you intend to provide services
+	is_lsp: bool,
+}
+
+/// The main interface into LSP functionality
+pub struct LSPManager<L: Deref, ES: Deref>
+where
+	L::Target: Logger,
+	ES::Target: EntropySource,
+{
+	logger: L,
+	pending_messages: Arc<Mutex<Vec<(PublicKey, LSPSMessage)>>>,
+	pending_events: Arc<Mutex<Vec<Event>>>,
+	request_id_to_method_map: Mutex<HashMap<String, String>>,
+	lsps0_message_handler: LSPS0MessageHandler<ES>,
+	config: LSPConfig,
+}
+
+impl<L: Deref, ES: Deref> LSPManager<L, ES>
+where
+	L::Target: Logger,
+	ES::Target: EntropySource,
+{
+	/// Constructor for the LSPManager
+	///
+	/// Sets up all required protocol message handlers based on configuration passed in
+	pub fn new(logger: L, entropy_source: ES, config: LSPConfig) -> Self {
+		let pending_messages = Arc::new(Mutex::new(vec![]));
+		let pending_events = Arc::new(Mutex::new(vec![]));
+
+		let lsps0_message_handler = LSPS0MessageHandler::new(
+			entropy_source,
+			vec![],
+			pending_messages.clone(),
+			pending_events.clone(),
+		);
+
+		Self {
+			logger,
+			pending_messages,
+			pending_events,
+			request_id_to_method_map: Mutex::new(HashMap::new()),
+			lsps0_message_handler,
+			config,
+		}
+	}
+
+	/// List the LSP protocols a node supports
+	///
+	/// Will receive the response via a ListProtocols event
```

Comment:
As said above, I'm not entirely sure we should expose this, and not just call and handle it internally upon first connection? If we want the user to be able to query the supported protocols per-counterparty we should probably store the response in an `Option` field as part of a `LSPSPeer` metadata struct or similar and provide a getter rather than exposing it via an event type.

```suggestion
	/// Will receive the response via a [`ListProtocols`] event
```